New dask + hf blog post

scharlottej13 opened the pull request:
Thanks for all the help so far @lhoestq @davanstrien! Another review would be much appreciated.

You're not quite done yet, though. Please make sure to follow this process (as documented [here](https://github.com/huggingface/blog/tree/main?tab=readme-ov-file#how-to-write-an-article-)):

- [x] Add an entry to [_blog.yml](https://github.com/huggingface/blog/blob/main/_blog.yml).
- [x] Add a thumbnail. There are no requirements here, but there is a [template](https://github.com/huggingface/blog?tab=readme-ov-file#how-to-get-a-nice-responsive-thumbnail) if it's helpful.
- [x] Check you use a short title and blog path.
- [x] Upload any additional assets (such as images) to the Documentation Images [repo](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog). This is to reduce bloat in the GitHub base repo when cloning and pulling. Try to have small images to avoid a slow or expensive user experience. Opened a PR here https://huggingface.co/datasets/huggingface/documentation-images/discussions/372
- [x] Add metadata (such as authors) to your `md` file. You can also specify `guest` or `org` for the authors.
- [x] Ensure the publication date is correct.

lhoestq commented on dask-nlp.md:
```diff
@@ -0,0 +1,240 @@
+- user: davanstrien
+---
+
+# Multi-GPU Text Classification with Hugging Face + Dask
```
Having "Text Classification" could imply that we do the training part as well (and there are so many classification training blog posts out there...)

maybe something more specific ? what do you think of this:

Scaling AI-based data processing with Hugging Face + Dask

scharlottej13 replied:
That's a great point, I like your suggestion!

lhoestq commented on dask-nlp.md:
````diff
@@ -0,0 +1,240 @@
+```python
+from transformers import pipeline
+
+def compute_scores(row):
````
would be cool to have the same code for pandas experimenting and for dask large scale inference

```suggestion
def compute_scores(texts):
```

(I believe it will be fasted as well, the current code was processing texts one by one)

lhoestq commented on dask-nlp.md:
```diff
@@ -0,0 +1,240 @@
+        device=device
+    )
+    results = pipe(
+        row["text"],
```
```suggestion
        texts.to_list(),
```

lhoestq commented on dask-nlp.md:
```diff
@@ -0,0 +1,240 @@
+
+df = df[:100]
+min_edu_score = 3
+df["edu-classifier-score"] = df.apply(compute_scores, axis=1)
```
```suggestion
df["edu-classifier-score"] = compute_scores(df.text)
```

lhoestq commented on dask-nlp.md:
````diff
@@ -0,0 +1,240 @@
+)
+```
+
+Next, we’ll use the HF [FineWeb-Edu classifier](https://huggingface.co/HuggingFaceFW/fineweb-edu-classifier) to judge the educational value of the web pages in our dataset. Web pages are ranked on a scale from 0 to 5, with 0 being not educational and 5 being highly educational. We can use pandas to do this on a smaller, 100-row subset of the data, which takes ~30 seconds on a M1 Mac with a GPU.
````
might need to update the " ~30 seconds" if it's really faster this way

scharlottej13 replied:
indeed! it now takes closer to 10 seconds (updated)

lhoestq commented on dask-nlp.md:
```diff
@@ -0,0 +1,240 @@
+def compute_scores(row):
+    import torch
+
+    # Select which hardware to use
+    if torch.cuda.is_available():
+    device = torch.device("cuda")        # NVIDIA GPU
+    elif torch.backends.mps.is_available():
+    device = torch.device("mps")         # Apple silicon GPU
+    else:
+    device = torch.device("cpu")         # CPU
```
```suggestion
    # Select which hardware to use
    if torch.cuda.is_available():
        device = torch.device("cuda")        # NVIDIA GPU
    elif torch.backends.mps.is_available():
        device = torch.device("mps")         # Apple silicon GPU
    else:
        device = torch.device("cpu")         # CPU
```

lhoestq commented on dask-nlp.md:
```diff
@@ -0,0 +1,240 @@
+def compute_scores(texts):
+    import torch
+
+    # Select which hardware to use
+    if torch.cuda.is_available():
+    device = torch.device("cuda")        # NVIDIA GPU
+    elif torch.backends.mps.is_available():
+    device = torch.device("mps")         # Apple silicon GPU
+    else:
+    device = torch.device("cpu")         # CPU
```
```suggestion
    # Select which hardware to use
    if torch.cuda.is_available():
        device = torch.device("cuda")        # NVIDIA GPU
    elif torch.backends.mps.is_available():
        device = torch.device("mps")         # Apple silicon GPU
    else:
        device = torch.device("cpu")         # CPU
```

lhoestq commented on dask-nlp.md:
````diff
@@ -0,0 +1,240 @@
+)
+```
+
+We’ll apply the `compute_scores` function for text classification in parallel on our Dask DataFrame using `map_partitions`, which is like a parallel version of a pandas `apply`, and applies our function in parallel on each pandas DataFrame in the larger Dask DataFrame. The `meta` argument is specific to Dask, and indicates the data structure (column names and data types) of the output.
````
`apply` transforms columns or rows, while map_partitions transforms dataframes IIUC

```suggestion
We’ll apply the `compute_scores` function for text classification in parallel on our Dask DataFrame using `map_partitions`, and applies our function in parallel on each pandas DataFrame in the larger Dask DataFrame. The `meta` argument is specific to Dask, and indicates the data structure (column names and data types) of the output.
```

lhoestq commented on dask-nlp.md:
```diff
@@ -0,0 +1,240 @@
+to_parquet(
+    df,
+    "hf://datasets/<your-hf-user>/<data-dir>"               # Replace with your HF user and directory
+    storage_options={"token": os.environ.get("HF_TOKEN")}
+)
```
the HF token is automatically loaded, you can remove this line

```suggestion
)
```

scharlottej13 replied:
oh cool!

lhoestq commented on _blog.yml:
```diff
@@ -4752,3 +4752,12 @@
+  thumbnail: /blog/assets/dask-nlp/thumbnail.png
+  date: October 8, 2024
+  tags:
+    - nlp
+    - guide
```
```suggestion
    - nlp
    - guide
    - datasets
```

lhoestq reviewed (commented):
Awesome ! I just left some minor comments but otherwise LGTM

scharlottej13 commented:
> Awesome ! I just left some minor comments but otherwise LGTM

Thanks for the review @lhoestq! I think I've addressed your comments, feel free to let me know if there's anything else.

pcuenca commented on dask-nlp.md:
```diff
@@ -0,0 +1,239 @@
+
+# Scaling AI-Based Data Processing with Hugging Face + Dask
+
+The Hugging Face platform has many datasets and pre-trained models that make using and training state-of-the-art machine learning models increasingly accessible. However, it can be hard to scale AI tasks because AI datasets are often large (100s to TBs) and using Hugging Face transformers for model inference can sometimes be computationally expensive.
```
```suggestion
The Hugging Face platform has many datasets and pre-trained models that make using and training state-of-the-art machine learning models increasingly accessible. However, it can be hard to scale AI tasks because AI datasets are often large (100s GBs to TBs) and using Hugging Face transformers for model inference can sometimes be computationally expensive.
```

pcuenca commented on dask-nlp.md:
```diff
@@ -0,0 +1,239 @@
+import dask.dataframe as dd
+
+df = dd.read_parquet(
+    "hf://datasets/HuggingFaceFW/fineweb/data/CC-MAIN-2024-10/*.parquet" # Load the full dataset lazily with Dask
```
Maybe put the comment above to avoid wrapped lines

pcuenca reviewed (approved):
Cool stuff!

scharlottej13 commented on dask-nlp.md:
Done!

I think I'm probably missing something obvious, but I wanted to also change https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog/dask-nlp to .../dask-scaling for clarity, but I can only see how to upload a new file (vs. changing the name of the directory).

Would you recommend deleting that folder + uploading a new one with the right name?

lhoestq replied:
I just renamed to dask-scaling in documentation-images :)

pcuenca commented on _blog.yml:
```diff
@@ -4752,3 +4752,13 @@
+  title: "Scaling AI-based Data Processing with Hugging Face + Dask"
+  author: scj13
+  thumbnail: /blog/assets/dask-scaling/thumbnail.png
+  date: October 8, 2024
```
Reminder to update

pcuenca commented:
Deferring to @lhoestq for final approval / merge.

davanstrien commented:
Looks great! Thanks for working on this @scharlottej13 🤗

lhoestq reviewed (approved):
Alright let's publish it 🙌

lhoestq commented:
blog post is out :)

https://huggingface.co/blog/dask-scaling

scharlottej13 commented:
Thanks everyone!
